### ISSUE TITLE
Improve stack integration.

### ISSUE THREAD

**Issue by aloiscochard** [aloiscochard/codex#46]
Currently the `stack` integration is minimal, it gives the same feature you would get from a normal cabal setup, we could do better, for example:

- Support dependency that are fetched from git (ref: https://github.com/aloiscochard/codex/issues/43)

Other ideas?

**Comment by aloiscochard** [comment]
For now codex only work with stack if the global GHC is used, we should find a way to configure cabal accordingly here:
https://github.com/aloiscochard/codex/blob/master/src/Codex/Project.hs#L91-L100

@snoyberg maybe you have some pointers/tips?

**Comment by snoyberg** [comment]
You can run relevant commands inside `stack exec` to get GHC in scope, not sure if that's what you're looking for though.

**Comment by aloiscochard** [comment]
@snoyberg the thing is: I'm using cabal as a library, and I would prefer to avoid people having to wrap codex call into stack... but I can understand that might be mandatory for people working with "embedded" GHCs.

@afcowie could you give that a try on your project? basically doing `stack codex update` should do the job (by that I mean `cabal` should find `ghc`), and so codex should run without printing any warning. (you might want to add `--force` to be sure it override the previously generated `codex.tags` file or simply remove it)

**Comment by snoyberg** [comment]
I don't understand the situation well enough right now. Using Cabal as a library shouldn't depend on GHC being on the PATH. What action is failing right now?

**Comment by aloiscochard** [comment]
@snoyberg so if you read those few lines: https://github.com/aloiscochard/codex/blob/master/src/Codex/Project.hs#L91-L100 you'll see that I'm basically doing what you suggested there: https://github.com/aloiscochard/codex/issues/39#issuecomment-142180391

But using the library function `configure` of `cabal` instead of calling the executable, I don't know the internals of cabal well enough, but it seems like at some point it requiered GHC, because @afcowie got this error: https://github.com/aloiscochard/codex/issues/39#issuecomment-150042118

The thing is, it works perfectly fine for me locally, because I have GHC installed globally on the system.

I suppose a way to workaround this problem would be to run `codex` using the `stack exec` "wrapper", but I was wondering if there is alternative, like calling `stack` in some way to get the path to GHC and pass that to the `configure` function of the cabal library...

Anyway, if it works with the wrapper, that can be a good enough solution for now.

Does that makes sense?

**Comment by snoyberg** [comment]
It makes sense. I didn't understand at first that `configure` was coming from the Cabal library, makes more sense now.

As I see it, you've got two additional options:

1. Ask `stack` for its modified PATH variable (`stack path --bin-path`), and set it globally on your process with `setEnv`
2. Similarly get the modified PATH, but then call out to `runghc Setup.hs` as a separate process using that modified PATH, avoiding a global modification inside your process

There should ideally be a way to modify Cabal's search path within your process, but I'm not sure how that's done (if it's at all possible).

**Comment by aloiscochard** [comment]
I suppose I'm fine with (1) as `codex` is a highly specialized tool and that shouldn't interfere with taggers (which are the only binaries used by `codex` from the env).

@snoyberg awesome thanks a lot!

**Comment by chreekat** [comment]
@aloiscochard note that one (common?) use case is having ghc installed globally, but operating within a package that uses a different version locally! It may be best to always rely on stack to pick a version of ghc when operating on a stackified project.

**Comment by aloiscochard** [comment]
@chreekat totally agree that's the plan.

It's quite simple to implement in fact, but I'm trying to find a way to test/debug it and sadly that seems impossible under NixOS.

I might setup a VM, or just do it in a branch and ask you folks for some help with testing if possible.

**Comment by chreekat** [comment]
Sure thing, I'll test branches!

**Comment by afcowie** [comment]
I tried `stack exec codex update` and it didn't seem to help (or rather, it came up with a different error message):

```
Configuring blah-0.1.3...
cabal: user error (Use of GHC's environment variable GHC_PACKAGE_PATH is
incompatible with Cabal. Use the flag --package-db to specify a package database
(it can be used multiple times).)
codex: *warning* falling back on dependency resolution using hackage
Updating blah-0.1.3
```

Not sure if that matters.

AfC

**Comment by aloiscochard** [comment]
@chreekat @afcowie sorry for late reply, I did not had a chance before to add that fix...

I added the potential fix in `feature/slack` (https://github.com/aloiscochard/codex/tree/feature/slack) could you guys give it a try and let me know if it solves your problem please?

Thanks in advance, I might find time to give it a try into a docker container... but in the meantime...

**Comment by chreekat** [comment]
Hey! Two things:

1. My problem *may* have been the fact that I was using fast-tags. With fast-tags, the feature/slack branch wasn't "fixed".
2. Once I installed and used hasktags, codex worked! It's amazing. :)

**Comment by afcowie** [comment]
Oh. *Not* to use `stack exec codex update` but to use `codex update`. Right.

Tried your feature/slack branch. `codex update` works cleanly now!

AfC

**Comment by aloiscochard** [comment]
The fix is release as part of `0.4.0.6` which is now on hackage! Thanks everyone for your help.

@chreekat if you are interested to get support for fast-tags you can file an issue here but tbh I'm not so sure it is worth it with `codex`, because `codex` already do some sophisticated caching and merging.
